### PR TITLE
VIDEO-7928: configure chrome simulcast layers

### PR DESCRIPTION
**Pull request by makarandp0**
This change configures simulcast layers as described in https://code.hq.twilio.com/client/sdk-frd/pull/96. 
- configure video and chrome screen-share track layers when adaptive simulcast is effective
- only set 'active: false' during default configuration
- apply default configuration after publisher hints are applied.
- after replaceTrack remove active flags for the layers that should be enabled.
- added bunch of tests

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [ ] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: lib/signaling/v2/peerconnection.js
Comment (charliesantos) [inline lib/signaling/v2/peerconnection.js:498]: Just curious, what is the reason why?

**Review by charliesantos — APPROVED** [review]

File: lib/signaling/v2/peerconnection.js
Comment (makarandp0) [reply lib/signaling/v2/peerconnection.js:498]: Because chrome does good job for screen share tracks - it keeps same dimensions but different frame rate for the simulcast layers. At one point we did configure the screen share track for chrome but found that it did not work as well (VIDEO-7000)

**Review by ceaglest — CHANGES_REQUESTED** [review]
This PR doesn't seem ready for review to me because the design is not completed for JavaScript and there is no agreed upon procedure to validate the results.

Please do not merge until we have agreed upon this behavior in the FRD and how to validate it in the QSDK test plan. Both of those items should be considered blockers for this PR.

**Comment by makarandp0** [comment]
> This PR doesn't seem ready for review to me because the design is not completed for JavaScript and there is no agreed upon procedure to validate the results.
> 
> Please do not merge until we have agreed upon this behavior in the FRD and how to validate it in the QSDK test plan. Both of those items should be considered blockers for this PR.

@ceaglest, I thought we had agreement on this, but now I have opened https://code.hq.twilio.com/client/sdk-frd/pull/96 to discuss this further. I also noticed that none of the adaptive simulcast frd changes are merged to master yet.

**Comment by makarandp0** [comment]
@ceaglest @manjeshbhargav, This PR now implements our decisions from discussion earlier. Please take a look.

File: lib/signaling/v2/peerconnection.js
Comment (makarandp0) [reply lib/signaling/v2/peerconnection.js]: Update: we decided to configure chrome screen share tracks as well. They will use different config than the camera tracks as described in  https://code.hq.twilio.com/client/sdk-frd/blob/jsdk_updates_for_adaptive_simulcast/video/adaptive-simulcast.md#configuring-simulcast

File: lib/signaling/v2/peerconnection.js
Comment (charliesantos) [inline lib/signaling/v2/peerconnection.js:573]: `_updateEncodings` will always return true correct? Maybe remove the return type here and just return true where we're calling it inside `_maybeUpdateEncodings`?

File: test/integration/spec/bandwidthprofile/publisherhints.js
Comment (charliesantos) [inline test/integration/spec/bandwidthprofile/publisherhints.js:65]: Please don't forget to remove.

**Review by charliesantos — CHANGES_REQUESTED** [review]

**Comment by makarandp0** [comment]
Thanks @charliesantos, I have incorporated your feedback.

**Review by charliesantos — APPROVED** [review]
Should we add changelog? Code generally looks good to me. But we should probably get +1 from @manjeshbhargav and/or @ceaglest

**Review by ceaglest — COMMENTED** [review]
The PR is ready for review now that the design is agreed upon, but I won't be able to review the JS implementation. Thank you @makarandp0 for the design updates and sorry if my previous comment was rude (it was made in a pretty stressful situation).

File: lib/signaling/v2/peerconnection.js
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js:14]: Any reason why this change was made. The previous require call was just one line.

File: lib/signaling/v2/peerconnection.js
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js:491]:
```suggestion
   * Whether adaptive simulcast is enabled.
   * @returns {boolean}
```

File: lib/signaling/v2/peerconnection.js
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js:493]: This can be a getter.
```suggestion
  get _isAdaptiveSimulcastEnabled() {
```

File: lib/signaling/v2/peerconnection.js
Comment (manjeshbhargav) [inline lib/signaling/v2/peerconnection.js:531]:
```suggestion
```

File: lib/util/sdp/simulcast.js
Comment (manjeshbhargav) [inline lib/util/sdp/simulcast.js:74]: Can you explain this change?

File: test/integration/spec/bandwidthprofile/publisherhints.js
Comment (manjeshbhargav) [inline test/integration/spec/bandwidthprofile/publisherhints.js:13]: This is not necessary.

File: lib/util/sdp/simulcast.js
Comment (makarandp0) [reply lib/util/sdp/simulcast.js:74]: Here instead of generating 3 random ssrcs, we generate one random ssrc and two others in sequence. I found this to be very useful to identify the layer indexes from the ssrc numbers. We are not making any assumption in the code (or tests) on this, but I think its useful.

File: lib/signaling/v2/peerconnection.js
Comment (makarandp0) [reply lib/signaling/v2/peerconnection.js:14]: This change was made to enable unit testing against different browsers. By importing `util` separately  we can  now mock guessBrowser [here](https://github.com/twilio/twilio-video.js/blob/8c575659750af12476891dc36071fae1a08c081a/test/unit/spec/signaling/v2/peerconnection.js#L87)

File: test/integration/spec/bandwidthprofile/publisherhints.js
Comment (makarandp0) [reply test/integration/spec/bandwidthprofile/publisherhints.js:13]: I think its usage does make it little bit more readable.

**Comment by makarandp0** [comment]
Thanks @manjeshbhargav, I have addressed your feedback. Can you please take another look?

File: lib/util/sdp/simulcast.js
Comment (makarandp0) [reply lib/util/sdp/simulcast.js]: now reverted

File: test/integration/spec/bandwidthprofile/publisherhints.js
Comment (manjeshbhargav) [reply test/integration/spec/bandwidthprofile/publisherhints.js:13]: I think this is not needed because the variable names indicate that the values are in milliseconds.

**Review by manjeshbhargav — APPROVED** [review]

**Pull request merged**
